albumentations/augmentations/functional.py

```diff
@@ -2041,3 +2041,48 @@ def glass_blur(img, sigma, max_delta, iterations, dxy, mode):
             x[h, w], x[h + dy, w + dx] = x[h + dy, w + dx], x[h, w]
 
     return np.clip(cv2.GaussianBlur(x / coef, sigmaX=sigma, ksize=(0, 0)), 0, 1) * coef
+
+
+def fourier_domain_adaptation(img, target_img, beta):
+    """
+    Fourier Domain Adaptation from https://github.com/YanchaoYang/FDA
+
+    Args:
+        img:  source image
+        target_img:  target image for domain adaptation
+        beta: coefficient from source paper
+
+    Returns:
+        transformed image
+
+    """
+    # get fft of both source and target
+    fft_src = np.fft.fft2(img.astype(np.float32), axes=(0, 1))
+    fft_trg = np.fft.fft2(target_img.astype(np.float32), axes=(0, 1))
+
+    # extract amplitude and phase of both fft-s
+    amplitude_src, phase_src = np.abs(fft_src), np.angle(fft_src)
+    amplitude_trg = np.abs(fft_trg)
+
+    # mutate the amplitude part of source with target
+    amplitude_src_shifted = np.fft.fftshift(amplitude_src, axes=(0, 1))
+    amplitude_trg_shifted = np.fft.fftshift(amplitude_trg, axes=(0, 1))
+
+    height, width, _ = amplitude_src_shifted.shape
+    border = (np.floor(np.min((height, width)) * beta)).astype(int)
+    center_x, center_y = np.floor(height / 2.0).astype(int), np.floor(width / 2.0).astype(int)
+
+    y1, y2 = center_y - border, center_y + border + 1
+    x1, x2 = center_x - border, center_x + border + 1
+
+    amplitude_src_shifted[y1:y2, x1:x2] = amplitude_trg_shifted[y1:y2, x1:x2]
+    amplitude_src_mutated = np.fft.ifftshift(amplitude_src_shifted, axes=(0, 1))
+
+    # mutated fft of source
+    fft_src_mutated = amplitude_src_mutated * np.exp(1j * phase_src)
+
+    # get mutated image
+    src_image_transformed = np.fft.ifft2(fft_src_mutated, axes=(0, 1))
```

Comment:
Maybe we can to reduce the copies of images?
In many rows we store image copy that we use only in next row.
It is not necessary, but would be great if we will use in place operations and reduce optional copies of images. Memory operations are one of the slowest operations for system.

Reply:
done.